lsp-types/src/Language/LSP/MetaModel/CodeGen.hs

```diff
@@ -0,0 +1,911 @@
+    -- valueToX c = CustomX c
+    -- or
+    -- valueToX _ = Nothing
+    fallThroughClause <- case customCon of
+        Just (TH.NormalC cn _) -> do
+          n <- lift $ TH.newName "n"
+          pure $ TH.Clause [TH.VarP n] (TH.NormalB (TH.ConE cn `TH.AppE` TH.VarE n)) []
+        Just _ -> fail "impossible: wrong custom constructor"
+        Nothing -> pure $ TH.Clause [TH.WildP] (TH.NormalB (TH.ConE 'Nothing)) []
+    let clauses = normalClauses ++ [fallThroughClause]
+        sd = TH.SigD fromBaseTypeN (TH.ArrowT `TH.AppT` ty `TH.AppT` (if custom then TH.ConT tn else TH.ConT ''Maybe `TH.AppT` TH.ConT tn))
+        fd = TH.FunD fromBaseTypeN clauses
+    pure [sd, fd]
+
+  -- JSON instances in terms of the conversion functions. These are straighftorward, but still good to generate since there are so many!
+  jsonDs <- lift
+    [d|
+     instance J.ToJSON $(TH.conT tn) where
+       toJSON e = J.toJSON ($(TH.varE toBaseTypeN) e)
+
+     instance J.FromJSON $(TH.conT tn) where
+       parseJSON val = do
+         v <- J.parseJSON val
+         $(
+           if custom
+           then [| pure $ $(TH.varE fromBaseTypeN) v |]
+           else [| case $(TH.varE fromBaseTypeN) v of { Just e -> pure e; Nothing -> fail $ "unrecognized enum value for " ++ enumNameString ++ ":" ++ show v; } |]
+          )
+      |]
+
+  pure $ [datad] ++ toBaseTypeD ++ fromBaseTypeD ++ jsonDs
+
+-- | Generate a type declaration corresponding to a type alias.
+genAlias :: TypeAlias -> CodeGenM TH.Dec
+genAlias TypeAlias{name, type_, documentation, since, proposed} = do
+  st <- asks symbolTable
+  tn <- case Map.lookup name st of
+    Just n  -> pure n
+    Nothing -> fail $ "no name for " ++ show name
+  addDocumentation tn documentation since proposed
+
+  rhs <- convertType type_
+
+  ctor <- makeConstrName Nothing name
+  -- We don't actually need JSONKey instances for all the base types, but it's close enough
+  let jsonKey = case type_ of { BaseType _ -> True; _ -> False }
+      newtypeDeriving = TH.DerivClause (Just TH.NewtypeStrategy)
+          ([TH.ConT ''J.ToJSON, TH.ConT ''J.FromJSON] ++ if jsonKey then [TH.ConT ''J.ToJSONKey, TH.ConT ''J.FromJSONKey] else [])
+  pure $ TH.NewtypeD [] tn [] Nothing (TH.NormalC ctor [(noBang, rhs)]) [stockDeriving, newtypeDeriving]
+  --pure $ TH.TySynD tn [] rhs
+
+(.=?) :: (J.KeyValue kv, J.ToJSON v) => J.Key -> Maybe v -> [kv]
+k .=? v = case v of
+  Just v' -> [k J..= v']
+  Nothing -> mempty
+
+-- | Generate a type declaration corresponding to a top-level named struct.
+genStruct :: Structure -> CodeGenM [TH.Dec]
+genStruct s@Structure{name, documentation, since, proposed} = do
+  let structName = name
+
+  st <- asks symbolTable
+  tn <- case Map.lookup name st of
+    Just n  -> pure n
+    Nothing -> fail $ "no name for " ++ show name
+  addDocumentation tn documentation since proposed
+
+  ctor <- lift $ makeToplevelName name
```

Comment:
can just use `tn` here? Or does it have to be a new name? or does it even matter?

Reply:
Then in `genMetaModel` no need to un-mangle names?

Reply:
nvm it breaks TH - got it